packages/math/base-elements.lua

```diff
@@ -0,0 +1,1209 @@
+    local constants = getMathMetrics().constants
+    local scaleDown = self:getScaleDown()
+    -- Determine relative Ys
+    if self.base then
+      self.base.relY = SILE.length(0)
+    end
+    if self.sub then
+      self.sub.relY = self.base.depth + maxLength(
+        self.sub.height + constants.lowerLimitGapMin * scaleDown,
+        constants.lowerLimitBaselineDropMin * scaleDown)
+    end
+    if self.sup then
+      self.sup.relY = 0 - self.base.height - maxLength(
+        constants.upperLimitGapMin * scaleDown + self.sup.depth,
+        constants.upperLimitBaselineRiseMin * scaleDown)
+    end
+    -- Determine relative Xs based on widest symbol
+    local widest, a, b
+    if self.sub and self.sub.width > self.base.width then
+      if self.sup and self.sub.width > self.sup.width then
+        widest = self.sub
+        a = self.base
+        b = self.sup
+      else
+        widest = self.sup
+        a = self.base
+        b = self.sub
+      end
+    else
+      if self.sup and self.base.width > self.sup.width then
+        widest = self.base
+        a = self.sub
+        b = self.sup
+      elseif self.sup then
+        widest = self.sup
+        a = self.base
+        b = self.sub
+      else
+        widest = self.base
+        a = self.sub
+        b = nil
+      end
+    end
+    widest.relX = SILE.length(0)
+    local c = widest.width / 2
+    if a then a.relX = c - a.width / 2 end
+    if b then b.relX = c - b.width / 2 end
+    local itCorr = self:calculateItalicsCorrection() * scaleDown
+    if self.sup then self.sup.relX = self.sup.relX + itCorr / 2 end
+    if self.sub then self.sub.relX = self.sub.relX - itCorr / 2 end
+    -- Determine width and height
+    self.width = maxLength(
+      self.base and self.base.width or 0,
+      maxLength(
+        self.sub and self.sub.width or 0,
+        self.sup and self.sup.width or 0
+      )
+    )
+    if self.sup then
+      self.height = 0 - self.sup.relY + self.sup.height
+    else
+      self.height = self.base and self.base.height or 0
+    end
+    if self.sub then
+      self.depth = self.sub.relY + self.sub.depth
+    else
+      self.depth = self.base and self.base.depth or 0
+    end
+  end,
+  output = function(self, x, y, line) end
+})
+
+-- terminal is the base class for leaf node
+elements.terminal = pl.class({
+  _base = elements.mbox,
+  _type = "Terminal",
+  styleChildren = function(self) end,
+  shape = function(self) end
+})
+
+elements.space = pl.class({
+  _base = elements.terminal,
+  _type = "Space",
+  __tostring = function(self)
+    return "space{w = "..self.width..", h = "..self.height..", d = "..self.depth.."}"
+  end,
+  _init = function(self, width, height, depth)
+    elements.terminal._init(self)
+    self.width = type(width) == "string" and SILE.length(width)
+      or width
+    self.height = type(height) == "string" and SILE.length(height)
+      or height
+    self.depth = type(depth) == "string" and SILE.length(depth)
+      or depth
```

Comment:
This doesn't seem quite right to me. First, shouldn't height and depth at least (if not width as well) be `SILE.measurement`s not `SILE.length`s? Would they ever have stretch/shrink values?

Secondarily, if you are going to accept multiple types and do type casting, you should use `SU.cast()` to do it not re-invent the wheel. These routines would let a mix-and-match bag of types though (lengths, measurements, numbers) which is less than ideal.

Lastly, any type of casting at all is going to slow things down just a little. Since math is a new thing here and you don't have technical debt to deal with, it might be best to be very strick with the types accepted and not do any casting at all. Either that or take advantage of the casting so you don't need to verbosely specify things as for example in `newStandardHspace()` on L1029-1052.

Reply:
I agree on your second and third points.

> This doesn't seem quite right to me. First, shouldn't height and depth at least (if not width as well) be SILE.measurements not SILE.lengths? Would they ever have stretch/shrink values?

This was already in Rui’s code, but I kept it, thinking that in the far future, it is not unthinkable to have vertical space adjustment in math. It also seemed simpler to have those three lengths of the same type.

Reply:
Fair enough. Lets just use `SU.cast()` here and be flexible about what it accepts. I think some of Rui's code predates the current type system which explains why this got a little messy.

Reply:
No, I agree with you about avoiding superfluous casts. I removed them in my latest commits.